dwh/indicator_lib.py

```diff
@@ -120,4 +120,221 @@ def calc_TX_NEW(patient_agg_obs:  pd.DataFrame, ARV_plan: str,
+    ART_plan_answer: The concept answer codes for START DRUG, CONTINUE REGIMEN, REFILLED
+    end_date_str: The string representation of the last date as 'YYYY-MM-DD'.
+  Returns:
+    The aggregated DataFrame with age/gender buckets.
+  """
+  end_date = datetime.today()
+  if end_date_str:
+    end_date = parse_date(end_date_str)
+  # Check for TB TREATMENT PLAN (if START/RELAPSE it means diagnosis was done)
+  tb_tx_df = patient_agg_obs[(patient_agg_obs['code'] == TB_TX_plan)].copy()
+  tb_tx_df['TB_TX'] = (tb_tx_df['last_value_code'].isin(TB_plan_answer))
+
+  # check if patient is on ART
+  art_tx_df = patient_agg_obs[(patient_agg_obs['code'] == ARV_plan)].copy()
+  art_tx_df['ART_TX'] = (art_tx_df['last_value_code'].isin(ART_plan_answer))
+
+  # check if patient was screened for TB
+  tb_screen_df = patient_agg_obs[(patient_agg_obs['code'] == TB_screening)].copy()
+  tb_screen_df['TB_screening'] = (tb_screen_df['last_value_code'].isin([YES_CODE]))
+
+  # join the 3 DF
+  temp_df=tb_tx_df.merge(art_tx_df[['patientId', 'ART_TX']],
+                         on='patientId').merge(tb_screen_df[['patientId', 'TB_screening']],
+                                               on='patientId')
+  # evaluate
+  temp_df['TX_TB'] = ((temp_df['ART_TX'] == True) & (temp_df['TB_TX'] == True) &
+                      (temp_df['TB_screening'] == True))
+  temp_df = _gen_counts_and_ratio(temp_df, end_date, 'TX_TB')
+  return temp_df
+
+def calc_TX_ML(patient_agg_obs:  pd.DataFrame, ARV_plan: str,
+                ARV_plan_answer: List[str], end_date_str: str = None) -> pd.DataFrame:
+  """Calculates TX_ML indicator with its corresponding disaggregations.
+
+  TX_ML indicator counts the number of  ART patients (who were on ART at the
+  beginning of the quarterly reporting period) and then had no clinical
+  contact since their last expected contact
+
+  Args:
+    patient_agg_obs: A DataFrame generated by `patient_query.find_patient_aggregates()`.
+    ARV_plan: The concept question code for ANTIRETROVIRAL PLAN
+    ARV_plan_answer: The concept answer codes for START DRUG, CONTINUE REGIMEN, REFILLED
+    end_date_str: The string representation of the last date as 'YYYY-MM-DD'.
+  Returns:
+    The aggregated DataFrame with age/gender buckets.
+  """
+  end_date = datetime.today()
+  if end_date_str:
+    end_date = parse_date(end_date_str)
+
+  art_tx_df = patient_agg_obs[(patient_agg_obs['code'] == ARV_plan)].copy()
+  art_tx_df['ART_TX'] = (art_tx_df['first_value_code'].isin(ARV_plan_answer))
+
+  # check if patient is HIV +
+  not_art_tx_df = patient_agg_obs[(patient_agg_obs['code'] == ARV_plan)].copy()
+  not_art_tx_df['NOT_ART_TX'] = (not_art_tx_df['last_value_code'].isnull())
+
+  # join the 3 DF
+  temp_df = art_tx_df.merge(not_art_tx_df[['patientId', 'NOT_ART_TX']], on='patientId')
+
+  # evaluate
+  temp_df['TX_ML'] = ((temp_df['ART_TX'] == True) & (temp_df['NOT_ART_TX'] == True))
```

Comment:
Hmm, `TX_ML` seems to be zero always because I think no row will satisfy this condition. On line 328, you are saying `ART_TX` is `True` if `first_value_code` is something non-null. So by definition `last_value_code` will be non-null too in that case. This does not seem right.

Reply:
I see, I'm curious how we can define this indicator. the tricky part is `..had no clinical
  contact since their last expected contact`

Reply:
I am not sure what exactly "had no clinical contact since their last expected contact" mean but maybe for each patient, we can find the `max_date` for all of their observations (as an indication of the last visit) and check if that is too old? Or maybe we should limit these `max_date`s to only certain codes?

Reply:
`last expected contact` is clinically referred to as RTC date, given to a patient to indicate his/her next scheduled appointment
 
![Screenshot 2021-08-09 at 9 48 30 AM](https://user-images.githubusercontent.com/10988005/128669056-fdb0efc4-898a-4499-81d6-762e3004022b.png)


Depending on the clinical status of a patient, this date ranges from 2 weeks to 5 months from last encounter date.  For example a stable patient who is virally suppressed will be given longer RTC date while unsuppressed patients will be given shorter RTC date

In the previous ETL implementation we addressed  `..had no clinical contact since their last expected contact`  by [lagging/leading RTC date](https://github.com/AMPATH/etl/blob/master/etl-scripts/calculated_tables/hiv_summary.sql#L199-L209) 

then write queries to evaluate all patients who had an RTC date (plus a buffer of +3months) falling within    the current reporting period and did not have any clinical encounters after their last encounter.

Reply:
Just checking the status of this PR; are we blocked on this? If yes, can we drop this indicator for now and finish the others. I think we decided that certain types (and locations?) of encounters need to be filtered out to be able to check if there was no clinical contact since expected RTC; is that correct @kimaina?

Reply:
I have removed this. I'll address in another ticket